src/System.Buffers/src/System/Buffers/ByteBufferPool.cs

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace System.Buffers
+{
+    public sealed class ByteBufferPool : IBufferPool<byte[]>
+    {
+        private const int NUMBER_OF_BUFFERS_IN_BUCKET = 50;
+        private int _maxBufferSize;
+        private BufferBucket[] _buckets;
+
+        // 2MB taken as the default since the average HTTP page is 1.9MB
+        // per http://httparchive.org/interesting.php, as of October 2015
+        public ByteBufferPool(int maxBufferSizeInBytes = 2048000)
+        {
+            int maxBuckets = SelectBucketIndex(maxBufferSizeInBytes);
+
+            _maxBufferSize = maxBufferSizeInBytes;
+            _buckets = new BufferBucket[maxBuckets + 1];
+
+            for (int i = 0; i < _buckets.Length; i++)
+            {
+                _buckets[i] = new BufferBucket(GetMaxSizeForBucket(i), NUMBER_OF_BUFFERS_IN_BUCKET);
```

Comment:
I went back and forth with the lazy initialization and I went with this way for a couple reasons:

1) Since the buffer pool is an instance class, if someone creates one then they know that they will need the buffers and it isn't initialized for them behind the scenes
2) In step with above, since we know someone needs one since it's created, I optimized for constant time (with the disclaimer of all buffers being full and needing to alloc one) Rent calls; the caller doesn't need to worry about if this Rent call will kick a Lazy initialization or not since the initialization cost is paid during initialization.
3) This is more a guess than the other two, but my guess on the usage pattern of a Buffer Pool is that it will probably be initialized during the program initialization with other book keeping (such as binding and listening to sockets or hitting the disk for I/O) so if the program needs a buffer pool then it's better (IMHO) to upfront the cost so you can have fast access when you really need perf later on doing important work.

Thoughts?